File: mnist-mlp/train.lua
cmd = torch.CmdLine();
cmd:text('Actively learn an MLP');
cmd:text('Strategy')
cmd:option('-method','mi','Active learning strategy. {random, mi, mi-decorr, entropy, confidence, confidence-2}');
cmd:option('-init',50,'Initial samples');
cmd:option('-inc',10,'How many examples to select each round');
cmd:option('-maxIter',200,'How many rounds');
cmd:option('-ntrials',30,'How many dropout samples');
cmd:option('-seed',1234,'random seed');
cmd:text('Dataset')
cmd:option('-data','../../dataset/mnist/data_train.t7','Dataset for training');
cmd:text('Model parameters');
cmd:option('-nlayers',3,'RNN layers');
cmd:option('-nh',1200,'RNN size');
cmd:option('-noutput',10,'RNN layers');
cmd:text('Optimization parameters');
cmd:option('-batch',256,'Batch size (Adjust base on GRAM and dataset size)');
cmd:option('-lr',1e-4,'Learning rate');
cmd:option('-decay',150,'Learning rate decay in epochs');
cmd:option('-l2',0.01,'L2 regularizer');
cmd:option('-epochs',300,'Epochs');
params=cmd:parse(arg);


require 'nn'
require 'cutorch'
require 'cunn' 
require 'nngraph'
require '../utils/optim_updates'

--initialize a set of seeds
torch.manualSeed(params.seed);
cutorch.manualSeedAll(params.seed);
seeds={};
for i=1,params.ntrials do
	seeds[i]={};
	seeds[i][1]=torch.random(1048576);
	seeds[i][2]=torch.random(1048576);
	seeds[i][3]=torch.random(1048576);
end

print('Loading dataset');
data=torch.load(params.data);
params.ninput=data['data']:size(2);
print('Building network');
net=require('../utils/MLP')(params.nlayers,params.ninput,params.nh,params.noutput,true);
criterion=nn.CrossEntropyCriterion():cuda();
logsoftmax=nn.LogSoftMax():cuda();
print('Initializing session');
paths.mkdir('sessions')
Session=require('../utils/session_manager');
session=Session:init('./sessions');
basedir=session:new(params);
paths.mkdir(paths.concat(basedir,'model'));
paths.mkdir(paths.concat(basedir,'mask'));

--Rewrite dropout forward for testing
function nn.Dropout:updateOutput(input)
	if self.inplace then
		self.output = input
	else
		self.output:resizeAs(input):copy(input)
	end
	if self.p > 0 then
		if self.train then
			--stock dropout during training
			self.noise:resizeAs(input)
			self.noise:bernoulli(1-self.p)
			if self.v2 then
				self.noise:div(1-self.p)
			end
			self.output:cmul(self.noise)
		else
			--correlated dropout during testing
			local tmp=input:size();
			tmp[1]=1;
			self.noise:resize(tmp)
			self.noise:bernoulli(1-self.p)
			if self.v2 then
				self.noise:div(1-self.p)
			end
			self.noise=self.noise:expandAs(input)
			self.output:cmul(self.noise)
		end
	end
	return self.output
end
--Data provider
function data:batch_train(batch_size,mask)
	local fv=torch.DoubleTensor(batch_size,self.data:size(2));
	local label=torch.DoubleTensor(batch_size);
	for i=1,batch_size do
		--randomly select an item
		local id=mask[torch.random(mask:size(1))];		
		label[i]=self.label[id];
		fv[i]=self.data[id];
	end
	return fv:cuda(),label:cuda();
end
function data:batch_eval(s,e)
	return self.data[{{s,e}}]:cuda(),self.label[{{s,e}}]:cuda();
end
--Network forward backward
function ForwardBackward(batch_size,mask)
	net.dw:zero();
	local fv,label=data:batch_train(batch_size,mask);
	local output=net.deploy:forward(fv);
	local f=criterion:forward(output,label);
	local doutput=criterion:backward(output,label);
	net.deploy:backward(fv,doutput);
	net.dw:add(params.l2/mask:size(1),net.w);
	return f;
end
--Network forward for evaluation
function Forward(s,e,t)
	local fv,label=data:batch_eval(s,e);
	torch.manualSeed(seeds[t][1]);
	cutorch.manualSeedAll(seeds[t][1]);
	local output=net.deploy:forward(fv);
	return logsoftmax:forward(output):clone();
end
--dem math
function logsumexp(s,dim)
	if dim==nil then
		local tmp = torch.max(s);
		return torch.log(torch.sum(torch.exp(s-tmp))) + tmp;
	else
		local tmp = torch.max(s,dim);
		return torch.log(torch.sum(torch.exp(s-tmp:expandAs(s)),dim)) + tmp;
	end
end
function logmeanexp(s,dim)
	if dim==nil then
		local tmp = torch.max(s);
		return torch.log(torch.mean(torch.exp(s-tmp))) + tmp;
	else
		local tmp = torch.max(s,dim);
		return torch.log(torch.mean(torch.exp(s-tmp:expandAs(s)),dim)) + tmp;
	end
end
function mi(s1,s2)
	local p_a_qi_theta_a=s1;
	local p_a_qi_theta_b=s2;
	local ntrials=s1:size(1);
	local noutput=s1:size(2);
	local pab=torch.repeatTensor(p_a_qi_theta_a:reshape(ntrials,noutput,1),1,1,noutput)+torch.repeatTensor(p_a_qi_theta_b:reshape(ntrials,1,noutput),1,noutput,1);
	pab=logmeanexp(pab,1):reshape(noutput,noutput);
	local pa=logmeanexp(p_a_qi_theta_a,1):reshape(noutput);
	local pb=logmeanexp(p_a_qi_theta_b,1):reshape(noutput);
	return -torch.sum(torch.cmul(pa,torch.exp(pa)))-torch.sum(torch.cmul(pb,torch.exp(pb)))+torch.sum(torch.cmul(pab,torch.exp(pab)));
end
function mst(d)
	local nvtx=d:size(1);
	local connected=d:clone():zero();
	local total=0;
	local es=d:reshape(nvtx*nvtx);
	local nu,ind=torch.sort(es,1,true);
	local j=1;
	for i=1,nvtx-1 do
		local found=0;
		while found==0 do
			x=(ind[j]-1)%nvtx+1;
			y=(ind[j]-x)/nvtx+1;
			j=j+1;
			if x<y then
				if connected[x][y]==0 then
					found=1;
					total=total+nu[j];
					connected[x][y]=1;
					connected[y][x]=1;
					connected[x][x]=1;
					connected[y][y]=1;
					for m=1,nvtx do
						for n=1,nvtx do
							if connected[x][m]==1 and connected[y][n]==1 then
								connected[m][n]=1;
								connected[x][n]=1;
								connected[y][m]=1;
								connected[n][m]=1;
								connected[n][x]=1;
								connected[m][y]=1;
							elseif connected[x][m]==1 then
								connected[y][m]=1;
								connected[m][y]=1;
							elseif connected[y][n]==1 then
								connected[x][n]=1;
								connected[n][x]=1;
							end
						end
					end
				end
			end
		end
	end
	return total;
end
log_file=paths.concat(basedir,string.format('log.txt',1));
function Log(msg)
	local f = io.open(log_file, "a")
	print(msg);
	f:write(msg..'\n');
	f:close()
end
print(params);

--Initialize mask
npts=data.data:size(1);
mask=torch.randperm(npts)[{{1,params.init}}]:clone():long();
torch.save(paths.concat(basedir,'mask',string.format('mask%d.t7',1)),mask);
--Active learning loop
scores=torch.DoubleTensor(npts,params.ntrials,params.noutput):fill(0);
local timer = torch.Timer();
for iter=1,params.maxIter do
	--train model
	--set up optimization
	Log(string.format('Iter %d',iter));
	niter_per_epoch=math.ceil(mask:size(1)/params.batch);
	local opt={};
	opt.maxIter=niter_per_epoch*params.epochs;
	opt.learningRate=params.lr;
	opt.running_avg=0;
	opt.decay=math.exp(math.log(0.1)/params.decay/niter_per_epoch);
	net.deploy:training();
	--optimization loop
	for i=1,opt.maxIter do
		local f=ForwardBackward(params.batch,mask);
		rmsprop(net.w,net.dw,opt);
		opt.learningRate=opt.learningRate*opt.decay;
		if i==1 then
			opt.running_avg=f;
		else
			opt.running_avg=opt.running_avg*0.95+f*0.05;
		end
	end
	torch.save(paths.concat(basedir,'model',string.format('model%d.t7',iter)),{net=net.net,trainloss=opt.running_avg});
	Log(string.format('\tt_model %f',timer:time().real));
	
	--compute score
	net.deploy:evaluate();
	for i=1,npts,params.batch do
		r=math.min(i+params.batch-1,npts);
		for t=1,params.ntrials do
			scores[{{i,r},t,{}}]=Forward(i,r,t):double();
		end
	end
	_,pred=torch.max(logmeanexp(scores,2):reshape(npts,params.noutput),2);
	acc=pred:eq(data.label:long()):sum()/pred:size(1);
	Log(string.format('\tt_score %f',timer:time().real));
	Log(string.format('\ttrainloss %f',opt.running_avg));
	Log(string.format('\ttrainacc %f',acc));
	
	--update mask
	entropy=torch.DoubleTensor(npts);
	expected_entropy=torch.DoubleTensor(npts);
	for i=1,npts do
		p_a_qi_theta=scores[{i,{},{}}]:cuda();
		p_a_qi=logmeanexp(p_a_qi_theta,1);
		entropy[i]=-torch.cmul(p_a_qi,torch.exp(p_a_qi)):sum();
		expected_entropy[i]=-torch.cmul(p_a_qi_theta,torch.exp(p_a_qi_theta)):mean(1):sum();
	end
	curiosity=entropy-expected_entropy;
	nu,ind=torch.sort(curiosity,1,true);
	sel={};
	nsel=math.min(params.inc,npts-mask:size(1));
	Log(string.format('\tmax_curiosity %f',curiosity:max()));
	Log(string.format('\ttop%d %f',params.inc,nu[{{1,nsel}}]:mean()));
	
	if params.method=='mi-decorr' then
		correlations={};
		ops=0;
		local curiosity_clone=curiosity:clone();
		curiosity_clone:index(1,mask):fill(-1e5);
		for i=1,nsel do
			candidate=-1;
			best_inc=-1;
			local nu,ind=torch.sort(curiosity_clone,1,true);
			for j=1,npts do
				max_correlation=0;
				for k=1,#sel do
					if correlations[(ind[j]-1)*npts+sel[k]]==nil then 
						tmp=mi(scores[{ind[j],{},{}}]:cuda(),scores[{sel[k],{},{}}]:cuda());
						ops=ops+1;
						correlations[(ind[j]-1)*npts+sel[k]]=tmp;
						correlations[(sel[k]-1)*npts+ind[j]]=tmp;
					end
					max_correlation=math.max(correlations[(ind[j]-1)*npts+sel[k]],max_correlation);
				end
				if nu[j]-max_correlation>best_inc then
					best_inc=nu[j]-max_correlation;
					candidate=ind[j];
				end
				if j<npts and best_inc>nu[j+1] then
					break;
				end
			end
			--print(string.format('Turn %d, selecting %d, original %f, gain %f, correlation %f',i,candidate,curiosity[candidate],best_inc,curiosity[candidate]-best_inc));
			table.insert(sel,candidate);
			curiosity_clone[candidate]=-1e5;
		end
		print(string.format('ops %d',ops));
		sel=torch.LongTensor(sel);
	elseif params.method=='confidence' then
		--per class, highest confidence
		if nsel%params.noutput~=0 then
			error('Cannot perform per-class selection');
		end
		local tmp=logmeanexp(scores,2):reshape(npts,params.noutput);
		for c=1,params.noutput do
			local nu,ind=torch.sort(tmp[{{},c}],1,true);
			local n=0;
			local t=1;
			while n<nsel/params.noutput do
				if c==1 and t==1 and mask:eq(ind[t]):sum()==0 then
					n=n+1;
					table.insert(sel,ind[t]);
				elseif torch.Tensor(sel):eq(ind[t]):sum()==0 and mask:eq(ind[t]):sum()==0 then
					n=n+1;
					table.insert(sel,ind[t]);
				end
				t=t+1;
			end
		end
		sel=torch.LongTensor(sel);
	elseif params.method=='mi' then
		tmp=curiosity:clone();
		tmp:index(1,mask):fill(-1e5);
		local nu,ind=torch.sort(tmp,1,true);
		i=1;
		while #sel<params.inc do
			if mask:eq(ind[i]):sum()==0 then
				if i>1 and torch.LongTensor(sel):eq(ind[i]):sum()==0 then
					table.insert(sel,ind[i]);
				elseif i==1 then
					table.insert(sel,ind[i]);
				end
			end
			i=i+1;
		end
		sel=torch.LongTensor(sel);
	elseif params.method=='entropy' then
		tmp=entropy:clone();
		tmp:index(1,mask):fill(-1e5);
		local nu,ind=torch.sort(tmp,1,true);
		i=1;
		while #sel<params.inc do
			if mask:eq(ind[i]):sum()==0 then
				if i>1 and torch.LongTensor(sel):eq(ind[i]):sum()==0 then
					table.insert(sel,ind[i]);
				elseif i==1 then
					table.insert(sel,ind[i]);
				end
			end
			i=i+1;
		end
		sel=torch.LongTensor(sel);
	elseif params.method=='random' then
		tmp=torch.rand(npts);
		tmp:index(1,mask):fill(-1e5);
		local nu,ind=torch.sort(tmp,1,true);
		i=1;
		while #sel<params.inc do
			if mask:eq(ind[i]):sum()==0 then
				if i>1 and torch.LongTensor(sel):eq(ind[i]):sum()==0 then
					table.insert(sel,ind[i]);
				elseif i==1 then
					table.insert(sel,ind[i]);
				end
			end
			i=i+1;
		end
		sel=torch.LongTensor(sel);
	else
		error('Unrecognized sample selection strategy.')
	end
	
	miab=torch.DoubleTensor(params.inc,params.inc);
	for i=1,params.inc do
	for j=1,params.inc do
		miab[i][j]=mi(scores[{sel[i],{},{}}]:cuda(),scores[{sel[j],{},{}}]:cuda());
	end
	end
	est_mi=mst(miab);
	Log(string.format('\tchosen %f',curiosity:index(1,sel):mean()));
	Log(string.format('\tchosen-corr %f',curiosity:index(1,sel):mean()-est_mi/params.inc));
	Log(string.format('\trandom %f',curiosity:mean()));
	
	mask=torch.cat(mask,sel,1);
	torch.save(paths.concat(basedir,'mask',string.format('mask%d.t7',iter+1)),mask);
	Log(string.format('\tt_mask %f',timer:time().real));
end

File: mnist-mlp/train.py
#Python 2/3 compatible headers
from __future__ import unicode_literals,division
from builtins import int
from builtins import range

#System packages
import torch
from torch.autograd import Variable,grad
import torch.nn.functional as F
import torch.nn as nn
import torch.optim as optim
import numpy
import scipy
import scipy.misc
import math
import time
import argparse
import sys
import re
import importlib

#Our libraries
import data.data as data
import util.math
import util.file
import util.session_manager as session_manager		#Manage runs


#Command line options
parser=argparse.ArgumentParser(description='Contextual distance distillation')
#	General
parser.add_argument('--seed', type=int, default=1)
parser.add_argument('--data_train', type=str, default='data/data_train.pt')
parser.add_argument('--data_test', type=str, default='data/data_test.pt')
#	Model
parser.add_argument('--arch', type=str, default='arch.mlp')
parser.add_argument('--nh', type=int, default=1024)
parser.add_argument('--nlayers', type=int, default=3)
parser.add_argument('--dropout', type=float, default=0.0)
#	Active selection
parser.add_argument('--method', type=str, default='entropy') #entropy, curiosity, goal
parser.add_argument('--init', type=int, default=50)
parser.add_argument('--rounds', type=int, default=100)
parser.add_argument('--inc', type=int, default=10)
parser.add_argument('--bnn', type=int, default=50)
parser.add_argument('--restart', action="store_true", default=False)
#	Optimization
parser.add_argument('--batch', type=int, default=256)
parser.add_argument('--lr', type=float, default=1e-3)
parser.add_argument('--epochs', type=int, default=30)
parser.add_argument('--times', type=int, default=3)
parser.add_argument('--ensemble', type=int, default=5)

params=parser.parse_args();
assert torch.cuda.is_available(),"You running on CPUs?"
torch.manual_seed(params.seed) #Fix random seed.
torch.cuda.manual_seed(params.seed);

params.argv=sys.argv;

#Load data
data_train=data.load(params.data_train);
data_test=data.load(params.data_test);
params.stuff=data_train.preprocess();
_=data_test.preprocess(params.stuff);

data_train.cuda();
data_test.cuda();
params.dof=data_train.dof;
params.noutput=data_train.noutput;

#Create Networks
arch=importlib.import_module(params.arch);
net=arch.new(params).cuda();
import util.pm as pm_proto
pms=[];
for i in range(0,params.ensemble):
	pms.append(pm_proto.dropout([p.data for p in net.parameters()]));

#Create session: keep everything in one folder
session=session_manager.Session(); #Create session
torch.save({'params':params},session.file('params.pt'));
tmp=vars(params);
tmp=dict([(k,tmp[k]) for k in tmp if not(k=='stuff')]);
print(tmp);
util.file.write_json(session.file('params.json'),tmp); #Write a human-readable parameter json
session.file('model','dummy');

#Randomly initialize active learning
subset=set(torch.randperm(data_train.size())[:params.init].tolist());
#Pre-allocate score matrix since it can get big
score_train=torch.Tensor(data_train.size(),params.bnn,params.noutput).cuda();
score_test=torch.Tensor(data_test.size(),params.bnn,params.noutput).cuda();
t0=time.time();
for iter_al in range(0,params.rounds):
	#Train ensemble of models
	#Optimization setup
	npts=len(subset);
	niter_per_epoch=max(math.ceil(npts/params.batch),5);
	subset_ind=torch.LongTensor(list(subset)).cuda();
	if params.restart:
		pms=[];
		for i in range(0,params.ensemble):
			pms.append(pm_proto.dropout([p.data for p in net.parameters()]));
	
	#Train a bunch of ensembles
	session.log('Iter %d, %d iter per epoch, time %f'%(iter_al,niter_per_epoch,time.time()-t0));
	net.train();
	for eid in range(0,params.ensemble):
		opt=optim.Adam(pms[eid].parameters(),lr=params.lr);
		for iter in range(1,niter_per_epoch*params.epochs+1):
			pms[eid].zero_grad();
			total_loss=0;
			for t in range(0,params.times):
				net.zero_grad();
				w=pms[eid]();
				pms[eid].impose(w,net.parameters())
				im,label=data_train.batch_sub(params.batch,subset_ind);
				im=Variable(im.cuda());
				label=Variable(label.cuda());
				score=net(im);
				loss=F.cross_entropy(score,label)/params.times;
				loss.backward();
				total_loss=total_loss+float(loss);
				pms[eid].propagate(w,net.parameters());
			opt.step();
			if iter%(niter_per_epoch*10)==0:
				session.log('\tensemble %d, epoch %.04f, loss %f, time %f'%(eid,iter/niter_per_epoch,total_loss,time.time()-t0));
	
	session.log('Iter %d, models trained, time %f'%(iter_al,time.time()-t0));
	#Randomly selects models for testing
	ensemble_ind=torch.LongTensor(params.bnn).random_(params.ensemble);
	seeds=torch.LongTensor(params.bnn).random_(1073741824);
	
	#Evaluate test accuracy
	net.eval();
	npts_test=data_test.size();
	test_acc=[];
	for i in range(0,npts_test,params.batch):
		r=min(npts_test,i+params.batch);
		for j in range(0,params.bnn):
			eid=int(ensemble_ind[j]);
			seed=float(seeds[j]);
			w=pms[eid](seed);
			pms[eid].impose(w,net.parameters());
			im,label=data_test.batch_eval(i,r);
			im=Variable(im.cuda());
			score_test[i:r,j,:]=F.log_softmax(net(im),dim=1).data;
		score=util.math.logmeanexp(score_test[i:r,:,:],dim=1);
		_,pred=score.max(dim=1);
		test_acc.append(label.eq(pred).float().sum()/npts_test);
	test_acc=sum(test_acc);
	session.log('Iter %d, test acc %f, time %f'%(iter_al,test_acc,time.time()-t0));
	
	#Evaluate scores for training pool examples
	net.eval();
	npts=data_train.size();
	for i in range(0,npts,params.batch):
		r=min(npts,i+params.batch);
		for j in range(0,params.bnn):
			eid=int(ensemble_ind[j]);
			seed=float(seeds[j]);
			w=pms[eid](seed);
			pms[eid].impose(w,net.parameters());
			im,label=data_train.batch_eval(i,r);
			im=Variable(im.cuda());
			score_train[i:r,j,:]=F.log_softmax(net(im),dim=1).data;
	session.log('Iter %d, train scores computed, time %f'%(iter_al,time.time()-t0));
	
	#Select new examples according to protocol
	if params.method=='entropy':
		logpy=util.math.logmeanexp(score_train,dim=1); #not very scalable but let's make do with it
		entropy=-(logpy*torch.exp(logpy)).sum(1);
		top_e,ind=entropy.sort(0,True);
		nsel=min(data_train.size()-len(subset),params.inc);
		n=0;
		total_entropy=0;
		for i in range(0,data_train.size()):
			if not(int(ind[i]) in subset):
				subset.add(ind[i]);
				n=n+1;
				total_entropy=total_entropy+float(top_e[i]);
				if n>=nsel:
					break;
		session.log('Iter %d, selected %d using entropy, entropy %f, time %f'%(iter_al,n,total_entropy,time.time()-t0));
	elif params.method=='curiosity':
		logpy=util.math.logmeanexp(score_train,dim=1,keepdim=True);
		model_param_mi=(torch.exp(score_train)*(score_train-logpy)).mean(1).sum(1);
		top_mi,ind=model_param_mi.sort(0,True);
		nsel=min(data_train.size()-len(subset),params.inc);
		n=0;
		total_gain=0;
		for i in range(0,data_train.size()):
			if not(int(ind[i]) in subset):
				subset.add(ind[i]);
				n=n+1;
				total_gain=total_gain+float(top_mi[i]);
				if n>=nsel:
					break;
		session.log('Iter %d, selected %d using curiosity, param information gain %f, time %f'%(iter_al,n,total_gain,time.time()-t0));
	elif params.method=='goal':
		#Compute average test vector
		avg_fim_test=[];
		for i in range(0,data_test.size(),params.batch):
			r=min(data_test.size(),params.batch);
			fim=score_test[i:r,:,:]-torch.logmeanexp(score_test[i:r,:,:],dim=1,keepdim=True);
			fim=torch.bmm(torch.exp(fim),torch.exp(score_test[i:r,:,:].clone().permute(0,2,1)))/params.bnn;
			avg_fim_test.append(fim.sum(0,keepdim=True)/data_test.size());
		avg_fim_test=torch.cat(avg_fim_test,dim=0);
		avg_fim_test=avg_fim_test.mean(0);
		#Compute best train-test matching
		train_test_mi=torch.Tensor(data_train.size()).cuda();
		for i in range(0,data_train.size(),params.batch):
			r=min(data_train.size(),params.batch);
			fim=score_train[i:r,:,:]-torch.logmeanexp(score_train[i:r,:,:],dim=1,keepdim=True);
			fim=torch.bmm(torch.exp(fim),torch.exp(score_train[i:r,:,:].clone().permute(0,2,1)))/params.bnn;
			train_test_mi[i:r]=(torch.mm(fim.view(r-i,-1),avg_fim_test.view(-1,1))-1)/2
		
		top_mi,ind=train_test_mi.sort(0,True);
		nsel=min(data_train.size()-len(subset),params.inc);
		n=0;
		total_gain=0;
		for i in range(0,data_train.size()):
			if not(int(ind[i]) in subset):
				subset.add(ind[i]);
				n=n+1;
				total_gain=total_gain+float(top_mi[i]);
				if n>=nsel:
					break;
		session.log('Iter %d, selected %d using goal-drive learning, testset information gain %f, time %f'%(iter_al,n,total_gain,time.time()-t0));
	
	torch.save({'pms':[pm.state_dict() for pm in pms],'subset':subset},session.file('model%02d.pt'%iter_al))
	